<a href="https://colab.research.google.com/github/Fellipe-A/Estudo_Python/blob/master/Python_zero4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<small><small><i>
Este notebook foi baseado nos notebooks disponíveis em [https://gitlab.erc.monash.edu.au/andrease/Python4Maths.git]
</i></small></small>

# Reaproveitando código

Uma outra maneira de utilizar um código múltiplas vezes é através de **procedimentos**.

Um procedimento é um conjunto de instruções agrupados em um bloco que recebem um nome e podem ser chamados quantas vezes for necessário.

Nos roteiros anteriores, você já vem usando procedimentos como ```input()```, ```print()``` e ```randint()```.

Você também pode criar seus próprios procedimentos usando o seguinte padrão:

```python
def nome():
    instrucao
    instrucao
    ...
```

O padrão acima pode ser lido como ```definir``` um procedimento chamado ```nome```, que executará o seguinte bloco de instruções.

As regras básicas para a **definição e uso de procedimentos** em Python são:

* O trecho de código delimitado por um procedimento é chamado bloco e deve ser identado.
* Um procedimento é chamado utilizando a notação ```nome()```.

In [0]:
def bom_dia():
    print("Bom dia!")

bom_dia()

Bom dia!


### Exercícios de fixação

1 - Crie um procedimento ```aleatorio()``` que imprima um inteiro aleatorio pertencente ao intervalo [1,60].

In [0]:
from random import randint
def aleatorio():
    aleatorio = randint(1,60)
    print(aleatorio)
aleatorio()


41


2 - Crie um procedimento mega_sena() que use o procedimento aleatorio() e imprima um jogo da mega sena.

In [0]:
def mega_sena():
  for i in range (6):
    numero = aleatorio()
mega_sena()

14
23
34
39
11
59


## Customizando a execução de um procedimento

Você pode configurar o comportamento de um procedimento utilizando **parâmetros de chamada** (também conhecidos como *argumentos*).

Não existe um limite para a quantidade de argumentos de um procedimento, mas você deve nomeá-los ao definir seu procedimento para poder utilizá-los como variáveis:

```python
def nome(arg1, arg2, ..., argN):
    instrucao
    instrucao
    ...
```


In [0]:
def info_pessoa(nome, idade, email):
    print("Nome:", nome)
    print("Idade:", idade)
    print("Email:", email)

info_pessoa("Leonardo Bezerra", "31", "leobezerra@imd.ufrn.br")

Nome: Leonardo Bezerra
Idade: 31
Email: leobezerra@imd.ufrn.br


### Exercícios de fixação

1 - Crie um procedimento ```media()``` que receba notas de 3 unidades de uma disciplina e calcule sua média.

In [0]:
def media(nota1,nota2,nota3):
  nota = (nota1 + nota2 + nota3)/3
  print(nota)
media(10,10,10)

10.0



2 - Crie um procedimento ```fatorial()``` que receba um inteiro N e imprima seu fatorial.

In [0]:

def fatorial(n):
  fact = 1
  for i in range (1,n+1):
    fact = fact*i
  print (fact)

fatorial(10)

3628800


## Interrompendo a execução de um procedimento

Assim como é possível interromper uma repetição usando o comando ```break```, é possível interromper a execução de um procedimento usando o comando ```return```.

In [0]:
def raiz_quadrada(n):
    if n < 0:
        return
    print("Raiz:", n ** 0.5)

raiz_quadrada(4)
raiz_quadrada(16)
raiz_quadrada(-1)

Raiz: 2.0
Raiz: 4.0


### Exercícios de fixação

1 - Crie um procedimento ```divisao_inteira()``` que imprima o resultado e o resto da divisão inteira de dois parâmetros inteiros. Se não houver resto, seu procedimento deverá imprimir apenas o resultado da divisão.

## Produzindo um valor a partir de um procedimento

Outro uso do comando ```return``` é permitir que um procedimento produza um resultado (retorne um valor).

Tanto é possível retornar valores diretamente, como usando as varíaveis associadas a esses valores.

In [0]:
def soma_n_primeiros(n):
    soma = 0
    for i in range(n + 1):
        soma += i
    return soma

print(soma_n_primeiros(3))
print(soma_n_primeiros(5))
print(soma_n_primeiros(10))

6
15
55


### Exercícios de fixação

1 - Crie um procedimento ```valor_absoluto()``` que retorne o valor absoluto de um parâmetro.

## Usando procedimentos definidos por outros

Um dos principais usos de procedimentos é a construção de **bibliotecas**, que são códigos criados por outras pessoas que você pode utilizar dentro do seu código.

> Indented block



Nesta disciplina, você poderá **importar** procedimentos de bibliotecas de duas formas:
- Importando procedimento(s) específico(s) de uma biblioteca
```python
from biblioteca import procedimento1, procedimento2, ..., procedimentoN
```
- Importando todos os procedimentos de uma biblioteca
```python
from biblioteca import *
```

### Exercícios de fixação

1 - Em roteiros anteriores, você usou a função ```randint()```. Qual biblioteca fornece esse procedimento?

Resposta: random




2 - Use os procedimentos ```sin()``` e ```floor()``` da biblioteca ```math``` para calcular o seno de um número e transformá-lo em um inteiro arredondando-o para baixo.

## Exercícios do URI

[2630](https://www.urionlinejudge.com.br/judge/pt/problems/view/2630) - Sistemas de computação gráfica representam imagens como um conjunto de pontos (pixels) formados por três componentes: vermelho, verde e azul.

Alguns algoritmos de processamento de imagem exigem um pré-processamento no qual é necessário transformar uma imagem colorida em uma imagem em tons de cinza. Esta conversão pode ser realizada de diversas maneiras, dependendo do resultado que se pretende obter.

Para preservar a percepção das cores básicas pelo olho humano, uma conversão apropriada seria tomar 30% da componente vermelha (R), 59% da componente verde (G) e 11% da componente azul (B). Em termos matemáticos,

P = 0,30R + 0,59G + 0,11B

Outras abordagens possíveis seriam determinar o valor de P através da média aritmética das três componentes ou atribuir a P os valores da maior ou da menor entre as três componentes.

Escreva um código Python que, dadas as componentes RGB de um pixel da imagem colorida, determine o valor do pixel P da imagem em tons de cinza correspondente, determinada a conversão a ser utilizada. Despreze a parte decimal do resultado, caso exista.

A entrada consiste em T (1 ≤ T ≤ 100) casos de teste, onde o valor de T é dado na primeira linha da entrada. Cada caso de teste é composto por quatro linhas: a primeira linha contém a conversão a ser utilizada: eye para a primeira abordagem descrita, mean para a média aritmética, max para o valor da maior componente e min para o valor da menor componente. As demais linhas contêm os valores R, G, B (0 ≤ R, G, B ≤ 255) do pixel da imagem colorida.

Para cada caso de testes dever ser impresso P, onde P é o nível de cinza do pixel da imagem em tons de cinza após a conversão do pixel da imagem colorida.

|.| Entrada | Saída |
|-|:----:|:-:|
| *Exemplo 1* | 3<br/>min<br/>35<br/>70<br/>35<br/>mean<br/>10<br/>74<br/>181<br/>eye<br/>23<br/>78<br/>197 | 35<br/>88<br/>74 | 
| *Exemplo 2* | 4<br/>min<br/>41<br/>27<br/>32<br/>min<br/>10<br/>21<br/>6<br/>max<br/>15<br/>41<br/>72<br/>max<br/>40<br/>21<br/>63 | 27<br/>6<br/>72<br/>41 | 
| *Exemplo 3* | 2<br/>eye<br/>7<br/>62<br/>200<br/>mean<br/>15<br/>49<br/>227 | 60<br/>97 | 

[2165](https://www.urionlinejudge.com.br/judge/pt/problems/view/2165) - O microblog Twitter é conhecido por limitar as postagens em 140 caracteres. Mais recentemente, alguns usuários como a @HBO_Brasil tiveram seus limites de caracteres dobrados para 280.

Escreva um código que confira se um texto vai caber em um tuíte. Cada caso de teste é iniciado com um inteiro informando a qual limite o usuário tem direito. Caso o tamanho do caracter exceda esse limite, seu código deverá imprimir a mensagem NO. Caso contrário, seu código deverá imprimir a mensagem OK.


|.| Entrada | Saída |
|-|----|:-:|
| *Exemplo 1* | 280<br/>Agora dá: Daenerys Targaryen, Filha da Tormenta, Não Queimada, Mãe de Dragões, Rainha de Mereen,<br/>Rainha dos Ândalos e dos Primeiros Homens, Quebradora de Correntes, Senhora dos Sete Reinos,<br/>Khaleesi dos Dothraki, a Primeira de Seu Nome. Descendente da Casa Targaryen. | OK | 
| *Exemplo 2* | 140<br/>Jon Snow. Ele é Rei do Norte. | OK | 
| *Exemplo 3* | 140<br/>Fish fingers last night! Looking forward to meeting with<br/>Prime Minister @Netanyahu shortly. (Was thinking of making a joke on<br/>his name – Netanya-WHO!? but @IvankaTrump said best not.)<br/>Peace in the Middle East would be a truly great legacy for ALL people! | NO | 




3 - Em exercícios de notebooks anteriores, você implementou códigos para calcular a contribuição mensal de um trabalhador para o INSS e seu IRPF (imposto de renda pessoa física).

Reuse seus códigos de forma modularizada para criar um código que leia o salário bruto de um trabalhador e imprima se salário líquido (descontados INSS e IRPF).

|.| Entrada | Saída |
|-|:-:|-|
| *Exemplo 1* | 1200.00 | 1104.00 | 
| *Exemplo 2* | 5485.00 | 4474.05 |
| *Exemplo 3* | 10431.00 | 8059.90 |